pkgs/build-support/docker/default.nix

```diff
@@ -886,4 +888,70 @@ rec {
         done;
       '' + extraCommands;
     });
+
+  # This function builds a docker image that behaves like a nix-shell
+  buildNixShellImage =
+    # Use the environment of this derivation
+    { name
+    , drv
+    , tag ? null
+    # Extra packages available inside the image
+    , extraPackages ? []
+    # Extra bash commands to execute when the image is ran
+    , extraInit ? ""
+    }:
+    # Remove once https://github.com/NixOS/nixpkgs/pull/77952 is merged
+    let
+
+      finalDrv = drv.overrideAttrs (old: {
+        nativeBuildInputs = old.nativeBuildInputs or [] ++ extraPackages;
```

Comment:
Why not just adding `extraPackages` in the contents list?

Reply:
Hm it's not the same thing: `contents` are added to the `/` path, but `extraPackages` are things that are added to `PATH` (and other env vars) later by stdenv's setup-hook. So by using `nativeBuildInputs` all kinds of dependencies should work, though now I'm debating whether that's even needed. Maybe it's needed for `man` support or so though.

Reply:
IIUC, this is to allow a user to add extra tooling in its container. I think, in other images, this use case is covered by the `contents` attribute, even if it is less convenient than what you are doing (because enviroment variables are not set). In order to have a coherent `dockerTools` API, I would prefer to expose the `dockerTools.buildLayeredImage` `contents` and `env` attributes: if the user needs extra tooling, it would be still possible to add packages with contents and set some environment variables with `env`. WDYT?

Reply:
The meaning is different. For consistency, users will expect `contents` to be available in the root, which is not happening for `extraPackages`.
Instead, it seems like `extraPackages` should be accepted by other `dockerTools` functions, to make adding store paths easier in those rare cases that it's needed.
However, first I'd rename it to `extraStorePaths` to avoid associations with `environment.systemPackages`.

Reply:
Thinking about this again, I feel like `extraPackages` should be removed, because people can just use `.overrideAttrs` on the derivation they pass in instead.

Similarly, I don't think an `env` is needed, because this can also be achieved by setting that in the derivation in multiple ways already:
```nix
{
  VAR = "VAL";
  shellHook = ''
    export VAR=VAL
  '';
}
```

I think the idea of passing through `contents` is a good one though, I'll change it as such